controllers/serviceexport/controller.go

```diff
@@ -0,0 +1,143 @@
+)
+
+// TO-DO (chenyu1): Update the API information when the design is finalized
+const (
+	// ServiceExport GVR information
+	svcExportGroup    = "mcs.networking.fleet.azure.io"
+	svcExportVersion  = "v1alpha1"
+	svcExportResource = "ServiceExport"
+
+	// ServiceImport GVR information
+	svcImportGroup    = "mcs.networking.fleet.azure.io"
+	svcImportVersion  = "v1alpha1"
+	svcImportResource = "ServiceImport"
+
+	// Min. and max. delay for workqueue rate limiters
+	minRetryDelay = time.Second * 5
+	maxRetryDelay = time.Second * 30
+)
+
+// ServiceExport GVR
+var svcExportGVR schema.GroupVersionResource = schema.GroupVersionResource{
+	Group:    svcExportGroup,
+	Version:  svcExportVersion,
+	Resource: svcExportResource,
+}
+
+// ServiceImport GVR
+var svcImportGVR schema.GroupVersionResource = schema.GroupVersionResource{
+	Group:    svcImportGroup,
+	Version:  svcImportVersion,
+	Resource: svcImportResource,
+}
+
+// Controller (ServiceExport controller) runs on a member cluster and syncs Services
+// (via ServiceExport and ServiceImport objects) and endpointSlices (via EndpointSliceExport objects)
+// with the hub cluster.
+type Controller struct {
+	// Unique ID of the member cluster in the fleet
+	memberClusterID string
+	// Clients for operating on the built-in and dynamic resources in the member cluster
+	memberKubeClient    clientset.Interface
+	memberDynamicClient dynamic.Interface
+	// Client for operating on the dynamic resources in the hub cluster
+	hubDynamicClient dynamic.Interface
+	// Informers for ServiceExport, Service, and EndpointSlice resources in the member cluster
+	memberSvcExportInformer     informers.GenericInformer
+	memberSvcInformer           coreinformers.ServiceInformer
+	memberEndpointSliceInformer discoveryinformers.EndpointSliceInformer
+	// Informer for ServiceImport resources in the hub cluster
+	hubSvcImportInformer informers.GenericInformer
+	// Workqueues for syncing Services with the hub cluster
+	memberSvcSyncWorkqueue workqueue.RateLimitingInterface
+	hubSvcSyncWorkqueue    workqueue.RateLimitingInterface
+	// Workqueue for syncing EndpointSlices with the hub cluster
+	memberEndpointSliceSyncWorkqueue workqueue.RateLimitingInterface
+	// Trackers that track if a resource has been successfully sync'd before and its last sync'd spec
+	svcSyncStatusTracker           SvcSyncStatusTracker
+	endpointSliceSyncStatusTracker EndpointSliceSyncStatusTracker
+	// Broadcaster and recorder for handling events
+	eventBroadcaster record.EventBroadcaster
+	eventRecorder    record.EventRecorder
+}
+
+func New(
+	memberClusterID string,
+	memberKubeClient clientset.Interface,
+	memberDynamicClient dynamic.Interface,
+	hubKubeClient clientset.Interface,
+	hubDynamicClient dynamic.Interface,
+	memberSharedInformerFactory informers.SharedInformerFactory,
+	memberDynamicInformerFactory dynamicinformer.DynamicSharedInformerFactory,
+	hubDynamicInformerFactory dynamicinformer.DynamicSharedInformerFactory,
+) (*Controller, error) {
```

Comment:
please break lines based on the semantics.

for example, putting the client in one line & putting the informerfactory in another line.

Reply:
Will do.